cpp/src/arrow/filesystem/gcsfs.h

```diff
@@ -0,0 +1,111 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/filesystem/filesystem.h"
+
+namespace arrow {
+namespace fs {
+class GCSFileSystem;
+struct GCSOptions;
+namespace internal {
+// TODO(ARROW-1231) - during development only tests should create a GCSFileSystem.
+//     Remove before declaring the feature complete.
+std::shared_ptr<GCSFileSystem> MakeGCSFileSystemForTest(const GCSOptions& options);
+}  // namespace internal
+
+struct ARROW_EXPORT GCSOptions {
+  std::string endpoint_override;
+  std::string scheme;
+
+  bool Equals(const GCSOptions& other) const;
+
+  /// \brief Initialize with default credentials provider chain
+  ///
+  /// This is recommended if you use the standard AWS environment variables
+  /// and/or configuration file.
+  static GCSOptions Defaults();
+
+  /// \brief Initialize with anonymous credentials.
+  ///
+  /// This will only let you access public buckets.
+  static GCSOptions Anonymous();
+};
+
+class ARROW_EXPORT GCSFileSystem : public FileSystem {
```

Comment:
small nit: GcsFileSystem? Also add docstring?

Reply:
I was wondering what was the style in this project. Fixed.